linera-service/src/linera-exporter/storage.rs

```diff
@@ -0,0 +1,428 @@
+    state_context: LogView<C, CryptoHash>,
+}
+
+impl<C> CanonicalState<C>
+where
+    C: Context + Send + Sync + 'static,
+{
+    fn new(cache_size: usize, state_context: LogView<C, CryptoHash>) -> Self {
+        let items_capacity =
+            (cache_size * 1024 * 1024) / (size_of::<usize>() + size_of::<CryptoHash>());
+
+        Self {
+            count: state_context.count(),
+            state_cache: Arc::new(FifoCache::new(items_capacity)),
+            state_updates_buffer: Arc::new(DashMap::new()),
+            state_context,
+        }
+    }
+
+    fn clone(&mut self) -> Result<Self, ViewError> {
+        Ok(Self {
+            count: self.count,
+            state_cache: self.state_cache.clone(),
+            state_updates_buffer: self.state_updates_buffer.clone(),
+            state_context: self.state_context.clone_unchecked()?,
+        })
+    }
+
+    async fn get(&self, index: usize) -> Result<CryptoHash, ExporterError> {
+        match self.state_cache.get_value_or_guard_async(&index).await {
+            Ok(value) => Ok(value),
+            Err(guard) => {
+                let hash = if let Some(entry) = self
+                    .state_updates_buffer
+                    .get(&index)
+                    .map(|entry| *entry.value())
+                {
+                    entry
+                } else {
+                    self.state_context
+                        .get(index)
+                        .await?
+                        .ok_or(ExporterError::UnprocessedBlock)?
+                };
+
+                let _ = guard.insert(hash);
+                Ok(hash)
+            }
+        }
+    }
+
+    async fn push(&mut self, value: CryptoHash) -> Result<(), ExporterError> {
+        let index = self.next_index();
+        let _ = self.state_updates_buffer.insert(index, value);
+        self.state_cache.insert(index, value);
+
+        Ok(())
+    }
+
+    fn flush(&mut self, batch: &mut Batch) -> Result<(), ExporterError> {
+        for (_, value) in self
+            .state_updates_buffer
+            .iter()
+            .map(|r| (*r.key(), *r.value()))
+            .collect::<BTreeMap<_, _>>()
+        {
+            self.state_context.push(value);
+        }
+
+        let _ = self.state_context.flush(batch)?;
+
+        self.state_updates_buffer.clear();
+        self.state_context.rollback();
+
+        Ok(())
+    }
+
+    fn next_index(&mut self) -> usize {
+        let next = self.count;
+        self.count += 1;
+        next
+    }
+}
+
+#[derive(Clone)]
+struct CacheWeighter<Q, V> {
+    key: PhantomData<Q>,
+    value: PhantomData<V>,
+}
+
+impl Weighter<BlobId, Arc<Blob>> for CacheWeighter<BlobId, Arc<Blob>> {
+    fn weight(&self, _key: &BlobId, val: &Arc<Blob>) -> u64 {
+        (size_of::<BlobId>()
+            + size_of::<Arc<Blob>>()
+            + 2 * size_of::<usize>()
+            + size_of::<Blob>()
+            + val.bytes().len()) as u64
```

Comment:
If we store `Arc<Blob>` in the cache then it's not true that every entry occupies the `val.bytes().len()` - it's `Arc`'d so it's only referencing the blob's content and not hold the bytes in the cache.